pkg/agent/controller/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,879 @@
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, nil, fmt.Errorf("packet file already exists. this may be due to an unexpected termination")
+	} else if os.IsNotExist(err) {
+		file, err = defaultFS.Create(filePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+//
+//	the agent on this Node will not perform the capture.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+
+	go func() error {
+		var err error
+		pcState := &packetCaptureState{name: name}
+		defer func() {
+			if err != nil {
+				c.cleanupPacketCapture(name)
+				klog.ErrorS(err, "PackageCapture failed", "name", name)
+				err := c.updatePacketCaptureStatus(name, pcState.numCapturedPackets, "", err)
+				if err != nil {
+					klog.ErrorS(err, "failed to update PacketCapture status")
+				}
+			}
+		}()
+		pc, err := c.packetCaptureLister.Get(name)
+		if err != nil {
+			return err
+		}
+
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil
+		}
+
+		klog.V(2).InfoS("Prepare capture on current node", "name", pc.Name, "device", *device)
+		packetSource, err := c.newPacketSourceFn(c, pc)
+		if err != nil {
+			klog.ErrorS(err, "failed to create packet source")
+			return err
+
+		}
+
+		pcState.maxNumCapturedPackets = pc.Spec.CaptureConfig.FirstN.Number
+		file, writer, err := getPacketFileAndWriter(pc.Name)
+		if err != nil {
+			return err
+		}
+
+		pcState.pcapngFile = file
+		pcState.pcapngWriter = writer
+		pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+		timeout := defaultTimeoutDuration
+		if pc.Spec.Timeout != nil {
+			timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+		}
+
+		err = c.performCapture(pcState, timeout, packetSource)
+		func() {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			c.captures[name] = "COMPLETED"
```

Comment:
what the usage of  'COMPLETED' info? I didn't see you check this kind of status.

Reply:
refactored to a  var. it's not used because the status update has been separated  to different functions:

1.  handling RUNNING/WAITING/UNKOWN -> before antrea-agent is doing the actual work
2. handling COMPLETED, which is for PacketCapture + FileUploaded. This function exist before we introduced the phase string.

I think it's not a ideal solution, which a single update function can handle all the status update, but there maybe too many conditions need to taken care of and it's will get too complex.